### DISCUSSION TITLE
Next.js App Router - client-side and server-side component with hydration boundary

### DISCUSSION THREAD

**Discussion by its-monotype** [TanStack/query#6267] [General]
I want only private pages like "profile" to fetch users using SSR so that public pages will stay static. I have a "header" component that renders on every page so I put it inside the root layout. But I want a prefetch query on certain pages like "profile" using SSR. The problem is that if I have a component with the query outside of the "hydration boundary" and a component inside the "hydration boundary" both are start running on the client.

```
src
├── api
│   └── get-user.tsx
├── app
│   ├── favicon.ico
│   ├── globals.css
│   ├── layout.tsx
│   ├── page.tsx
│   └── profile
│       ├── page.tsx
│       └── profile.tsx
└── components
    ├── header.tsx
    └── providers.tsx
```

```ts
export const getUser = async () => {
  await new Promise((resolve) => setTimeout(resolve, 1000));

  const res = await fetch('https://jsonplaceholder.typicode.com/users/1', {
    cache: 'no-store',
  });
  const data = await res.json();
  return data;
};
```

Root layout:
```tsx
import { Header } from '@/components/header';
import { Providers } from '@/components/providers';
import type { Metadata } from 'next';
import { Inter } from 'next/font/google';
import './globals.css';

const inter = Inter({ subsets: ['latin'] });

export const metadata: Metadata = {
  title: 'Create Next App',
  description: 'Generated by create next app',
};

export default async function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="en">
      <body className={inter.className}>
        <Providers>
          <Header />
          {children}
        </Providers>
      </body>
    </html>
  );
}
```

Providers:
```tsx
'use client';

import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ReactQueryDevtools } from '@tanstack/react-query-devtools';
import { useState } from 'react';

export const Providers = ({ children }: { children: React.ReactNode }) => {
  const [queryClient] = useState(
    () =>
      new QueryClient({
        defaultOptions: {
          queries: {
            // With SSR, we usually want to set some default staleTime
            // above 0 to avoid refetching immediately on the client
            staleTime: 60 * 1000,
          },
        },
      }),
  );

  return (
    <QueryClientProvider client={queryClient}>
      {children}
      <ReactQueryDevtools initialIsOpen={true} />
    </QueryClientProvider>
  );
};

```

Header:
```tsx
'use client';

import { getUser } from '@/api/get-user';
import { useQuery } from '@tanstack/react-query';

export const Header = () => {
  const { data: user } = useQuery({
    queryKey: ['user'],
    queryFn: getUser,
  });

  return <div className="text-white">{user?.name}</div>;
};

```

---

Profile page:
```tsx
import { getUser } from '@/api/get-user';
import {
  HydrationBoundary,
  QueryClient,
  dehydrate,
} from '@tanstack/react-query';
import { Profile } from './profile';

export default async function ProfilePage() {
  const queryClient = new QueryClient();
  await queryClient.prefetchQuery({
    queryKey: ['user'],
    queryFn: getUser,
  });

  return (
    <HydrationBoundary state={dehydrate(queryClient)}>
      <Profile />
    </HydrationBoundary>
  );
}
```

Profile:
```tsx
'use client';

import { getUser } from '@/api/get-user';
import { useQuery } from '@tanstack/react-query';

export const Profile = () => {
  const { data: user } = useQuery({
    queryKey: ['user'],
    queryFn: getUser,
  });

  return <div>Welcome to profile {user?.name}!</div>;
};
```

Everything works fine if I put "header" separately on each page like that:
```diff
export default async function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="en">
      <body className={inter.className}>
        <Providers>
-         <Header />
          {children}
          ...
```
```diff
export default async function HomePage() {
  return (
    <>
+   <Header/>
    <main className="flex min-h-screen flex-col items-center justify-between p-24">
      <div className="z-10 max-w-5xl w-full items-center justify-between font-mono text-sm lg:flex">
        <p className="fixed left-0 top-0 flex w-full justify-center border-b border-gray-300 bg-gradient-to-b from-zinc-200 pb-6 pt-8 backdrop-blur-2xl dark:border-neutral-800 dark:bg-zinc-800/30 dark:from-inherit lg:static lg:w-auto  lg:rounded-xl lg:border lg:bg-gray-200 lg:p-4 lg:dark:bg-zinc-800/30">
          Get started by editing&nbsp;
          ...
```
```diff
export default async function ProfilePage() {
  const queryClient = new QueryClient();
  await queryClient.prefetchQuery({
    queryKey: ['user'],
    queryFn: getUser,
  });

  return (
    <HydrationBoundary state={dehydrate(queryClient)}>
+     <Header />
      <Profile />
    </HydrationBoundary>
  );
}
```

the problem is that the header was using a "user" query and was outside HydrationBoundary. Is there any way to work around it? It feels tedious/wrong to add the "header" to each page or page layout if I need it on every page. I'd be grateful for your help!

**Comment by its-monotype** [comment]
So far, the best option that I have come up with is [this](https://github.com/its-monotype/next-app-dir-react-query)
If you have an idea on how to improve/simplify it, I would be grateful.

**Comment by its-monotype** [comment]
So my case here can be represented simpler like that: we have a server component that renders the client component that uses under-the-hood query, directly or wrapped with a server component that prefetches the query and wraps it inside HydrationBoundary like that:

```tsx
export default async function ServerComponent() {
  <User /> // client component
  <UserHydrated /> // same component, but wrapped in a hydration boundary
}
```

User:
```tsx
'use client';
import { getUser } from '@/api/get-user';
import { useQuery } from '@tanstack/react-query';

export const User = () => {
  const { data: user } = useQuery({
    queryKey: ['user'],
    queryFn: getUser,
  });

  return (
    <h1 className="text-6xl font-bold text-center mb-8">
      Bro {user?.name} you are logged in!
    </h1>
  );
};
```

UserHydrated:
```tsx
import { getUser } from '@/api/get-user';
import {
  HydrationBoundary,
  QueryClient,
  dehydrate,
} from '@tanstack/react-query';
import { User } from './user';

export const UserHydrated = async () => {
  const queryClient = new QueryClient();
  await queryClient.prefetchQuery({
    queryKey: ['user'],
    queryFn: getUser,
  });

  return (
    <HydrationBoundary state={dehydrate(queryClient)}>
      <User />
    </HydrationBoundary>
  );
};
```  

And with that setup, SSR does not work, and both components render on the client. But if I comment out raw client component then SSR works fine for UserHydrated. Is that how things should work that you must wrap all prefetched queries with HydrationBoundary or it's possible to use SSR for both when HydrationBoundary was used near the client component, or maybe some work around it?

**Reply by TkDodo** [reply]
I think you can if you re-use the queryClient instead of creating a new one: https://tanstack.com/query/latest/docs/react/guides/advanced-ssr#alternative-use-a-single-queryclient-for-prefetching

**Reply by its-monotype** [reply]
Works if I reuse the same queryClient for both prefetching and inside client providers but now page started flashing white for a second before load

**Reply by TkDodo** [reply]
the docs suggest to reuse the same client for prefetching only; the client in the browser needs to be a separate one

**Reply by its-monotype** [reply]
Yes, I tried to do it as in the docs (same queryClient for prefetching, btw I'm prefetching query only in one place), but the problem still exists, so I also tried to reuse it for both client and server, but it definitely doesn’t work

**Reply by its-monotype** [reply]
here is the code: https://github.com/its-monotype/next-app-dir-react-query/tree/reuse-client

**Reply by TkDodo** [reply]
plase try to make it work in codesandbox: 

https://codesandbox.io/p/github/its-monotype/next-app-dir-react-query/main

**Reply by its-monotype** [reply]
please try this one https://codesandbox.io/p/github/its-monotype/next-14-rq

**Reply by its-monotype** [reply]
@TkDodo did you manage to open it?

**Reply by its-monotype** [reply]
Here is a video

https://github.com/TanStack/query/assets/79363260/c028d95c-b25e-4ac5-9009-a1d4c8aa0a24

**Reply by TkDodo** [reply]
I'm not sure what you'd expect to happen. What happens is that:

- `Header` renders on the client, because it's a client component. There's no server component "above it" that does prefetching.
- `Profile` itself first prefetches on the server; while that happens react suspends. That's why you see the header, but no profile text. If you'd add a `loading.tsx`, you would see that.
- then, when profile is done fetching on the server, it hydrates data into the cache on the client and finishes rendering the component

what would you expect to happen in this situation?

> SSR does not work

not sure what "does not work" exactly, but you can clearly see that ssr "works" by adding a log statement to `getUser`:

```
console.log('fetching on', typeof window === 'undefined' ? 'server' : 'client')
```
